text/3808-register-tool.md

```diff
@@ -0,0 +1,289 @@
+- Feature Name: `register_tool`
+- Start Date: 2025-03-22
+- RFC PR: [#3803](https://github.com/rust-lang/rfcs/pull/3808)
+- Rust Issue: [rust-lang/rust#66079](https://github.com/rust-lang/rust/issues/66079)
+
+# Summary
+[summary]: #summary
+
+This RFC adds two new attributes:
+- `#![register_lint_tool(tool_name)]` allows controlling namespaced lints with `#[warn(tool_name::lint_name)]`
+- `#![register_attribute_tool(tool_name)]` allows using tool names in [inert attributes][inert] with `#[tool_name::attribute_name(token_tree)]`.
```

Comment:
May I suggest we use `#![register_tool_{attribute|lint}` instead? I was also wondering if we could have a `#[register_tool]` which would add the tool to both namespaces, the attribute and lint.

Reply:
I worry that `register_tool_{attribute|lint}` will imply we are registering _individual_ attributes or _individual_ lints, when that is not the case. If someone sees the attribute `#![register_tool_lint(...)]`, I worry they'll do this:

```rust
#![register_tool_lint(bevy::panicking_methods)]
#![register_tool_lint(bevy::missing_reflect)]
#![register_tool_lint(bevy::duplicate_bevy_dependencies)]
// ...
```

I think the name `register_lint_tool` much better signals that it's for registering a _tool_, not a _lint_ or _attribute_.

Reply:
Gotcha... That makes sense. Thanks

Reply:
unless we allow nested tool name like `com::example::tool`[^1], I don't think `#![register_tool_lint(bevy::panicking_methods)]` will compile because the tool name can only be an `$:ident` not a `$:path`, so it is actually not an issue.

[^1]: `#![warn(com::example::tool::lint)]` is interpreted as having a tool name `com`.

Reply:
sure, but why not avoid the confusion in the first place?

Reply:
> sure, but why not avoid the confusion in the first place?

IMO `#![register_tool(clippy)]` is also confusing, it sounds like it will make `rustc` run the "tool" `clippy` by "registering" it to the compilation process. The standard reference refers to this as "tool *name*" rather than just "tool". I think `#![register_tool_name(clippy)]` would be clear, but `#![register_attribute_tool_name(clippy)]` is quite verbose.

Reply:
That's not what "register" means. I think there is *some* amount of confusion possible, but I don't think it's likely enough that we need to make all the names longer. If people are adding these attributes, it's because the tool's docs told them to, and the docs should also say how to run the tool. If people are reading these attributes in existing code, then, well, I don't think it's that much more confusing than any other attribute in the language? `#[inline]` especially does not mean what most people think it means.